compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -453,7 +453,17 @@ impl<'p, 'tcx> Matrix<'p, 'tcx> {
     /// expands it.
     fn push(&mut self, row: PatStack<'p, 'tcx>) {
         if !row.is_empty() && row.head().is_or_pat() {
-            self.patterns.extend(row.expand_or_pat());
+            let pats = row.expand_or_pat();
```

Comment:
I couldn't make that one recursive but made a new function that is recursive.

Reply:
Either the other use of expand_or_Pat is dead code now or it's still got that issue. Check whether it's dead by panicking inside of it instead of doing any logic

Reply:
It can't be dead code because line 835, it's the only reason I couldn't replace the function altogether

Reply:
That call may be dead after your change is what I'm thinking. Since you now expand all the nested `Or` patterns, line 835 may not encounter `Or` patterns anymore.

Reply:
I did as you said, it panicked. Newly added function is only called via `fn push` and that is called via line 846.

Reply:
I don't like that we now have one function that does the recursive work and one that doesn't. On what kind of matches did it panic? Should we call the new function at that site to reduce duplication and make it more general?

Reply:
that is suspicious indeed and makes me confident that we should definitely make them work the same and figure out why they are problematic

Reply:
I think I understand now why the compiler gives "unreachable pattern" err, to reach those places I need to access `is_under_guard` for each recursive call, but also push to the matrix has to happen after the call `is_useful` to trigger `if v.is_empty()` where the matrix has to be empty. Even if I throw all those out of the window I need to call `is_useful` for each `v` 🤷‍♂️